src/literal.h

```diff
@@ -33,52 +33,69 @@ class Literals;
   union {
+    // i32 or f32 bits
     int32_t i32;
+    // i64 or f64 bits
     int64_t i64;
+    // v128 bytes
     uint8_t v128[16];
-    Name func; // function name for funcref
-    std::unique_ptr<ExceptionPackage> exn;
+    // funcref function name. not set indicates `null`.
+    Name func;
+    // exnref package. `nullptr` indicates `null`. we use a pointer here to
+    // limit storage size, with the lifetime of the data managed by the literal.
+    ExceptionPackage* exn;
+    // literals of other reference types can only be `null`
   };
 
 public:
-  Type type;
-
-public:
-  Literal() : v128(), type(Type::none) {}
-  explicit Literal(Type type) : v128(), type(type) {
-    assert(type != Type::unreachable);
+  Literal() : type(Type::none), v128() {}
+  explicit Literal(Type::BasicID typeId) : Literal(Type(typeId)) {}
+  explicit Literal(Type type) : type(type), v128() {
+    assert(type != Type::unreachable && (!type.isRef() || type.isNullable()));
   }
-  explicit Literal(int32_t init) : i32(init), type(Type::i32) {}
-  explicit Literal(uint32_t init) : i32(init), type(Type::i32) {}
-  explicit Literal(int64_t init) : i64(init), type(Type::i64) {}
-  explicit Literal(uint64_t init) : i64(init), type(Type::i64) {}
+  explicit Literal(int32_t init) : type(Type::i32), i32(init) {}
+  explicit Literal(uint32_t init) : type(Type::i32), i32(init) {}
+  explicit Literal(int64_t init) : type(Type::i64), i64(init) {}
+  explicit Literal(uint64_t init) : type(Type::i64), i64(init) {}
   explicit Literal(float init)
-    : i32(bit_cast<int32_t>(init)), type(Type::f32) {}
+    : type(Type::f32), i32(bit_cast<int32_t>(init)) {}
   explicit Literal(double init)
-    : i64(bit_cast<int64_t>(init)), type(Type::f64) {}
+    : type(Type::f64), i64(bit_cast<int64_t>(init)) {}
   // v128 literal from bytes
   explicit Literal(const uint8_t init[16]);
   // v128 literal from lane value literals
   explicit Literal(const std::array<Literal, 16>&);
   explicit Literal(const std::array<Literal, 8>&);
   explicit Literal(const std::array<Literal, 4>&);
   explicit Literal(const std::array<Literal, 2>&);
-  explicit Literal(Name func) : func(func), type(Type::funcref) {}
-  explicit Literal(std::unique_ptr<ExceptionPackage> exn)
-    : exn(std::move(exn)), type(Type::exnref) {}
+  explicit Literal(Name func) : type(Type::funcref), func(func) {}
+  explicit Literal(ExceptionPackage& exn_);
   Literal(const Literal& other);
   Literal& operator=(const Literal& other);
-  ~Literal() {
-    if (type == Type::exnref) {
-      exn.~unique_ptr();
-    }
-  }
+  ~Literal();
 
   bool isConcrete() const { return type != Type::none; }
   bool isNone() const { return type == Type::none; }
+  bool isNull() const {
+    if (type.isNullable()) {
+      if (type.isFunction()) {
+        return func.isNull();
+      }
+      if (type.isException()) {
+        return exn == nullptr;
+      }
+      return true;
+    }
+    return false;
+  }
```

Comment:
With `nullref` removed, literals of any nullable reference type can now be `null`, which is the default unless `func` or `exn` are set.